src/3d/qgs3dexportobject.cpp

```diff
@@ -0,0 +1,159 @@
+/***************************************************************************
+  Qgs3DExportObject.cpp
+  --------------------------------------
+  Date                 : June 2020
+  Copyright            : (C) 2020 by Belgacem Nedjima
+  Email                : gb underscore nedjima at esi dot dz
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgs3dexportobject.h"
+
+#include <QVector3D>
+#include <QDebug>
+#include <QDir>
+#include <QImage>
+
+Qgs3DExportObject::Qgs3DExportObject( const QString &name, const QString &parentName, QObject *parent )
+  : QObject( parent )
+  , mName( name )
+  , mParentName( parentName )
+  , mSmoothEdges( false )
+{
+}
+
+void Qgs3DExportObject::setupPositionCoordinates( const QVector<float> &positionsBuffer, float scale, const QVector3D translation )
+{
+  for ( int i = 0; i < positionsBuffer.size(); i += 3 )
+  {
+    for ( int j = 0; j < 3; ++j )
+    {
+      mVertxPosition << positionsBuffer[i + j] * scale + translation[j];
+    }
+  }
+
+  for ( int i = 0; i < positionsBuffer.size() / 3; ++i )
+  {
+    mIndexes << i + 1;
+  }
+}
+
+void Qgs3DExportObject::setupPositionCoordinates( const QVector<float> &positionsBuffer, const QVector<unsigned int> &faceIndex, float scale, const QVector3D translation )
+{
+  // TODO: delete vertices that are not used
```

Comment:
Does it happen that we have vertices that do not get used?

Reply:
I don't think that happens actually, I wrote this when I was looking for why invalid vertices are returned. I think that could happen if the height map is not square and we end up storing vertex coordinates for invalid faces even though we don't actually use it.
This code here detects invalid faces:
```
    if ( faceIndex[i] == faceIndex[i + 1] && faceIndex[i + 1] == faceIndex[i + 2] ) continue;
```